doConcurrentLayoutAndUpdate to speedup Score::update() (Not a PR...just sharing for thoughts)

ericfont opened the pull request:
I was pondering how much potential there was to speed up very large scores (see "Score Editing Slow" issue https://musescore.org/en/node/280952).  While familiarizing myself with the update code, I found that the big scores with lots of parts get bogged down even when just changing pitch of one note because they have to regenerate layout and update of the affected system for every single part that has been created.  That is in addition to the very large main score.

I notice that the layout and update of each individual part is independent* of eachother and the main score.  And turns out that QtConcurrent has nice high-level functions for map-reduce.  So basically what I've done is rearrange Score::update() code to be able to deal with each score independently, and spun off each individual score's independent processing into `Score::doConcurrentLayoutAndUpdate()`, and then in the main Score::update used `QtConcurrent::blockingMap(scores, [](Score* s){ s->doConcurrentLayoutAndUpdate(); } );` to concurrently execute the layout & update for every score in the MasterScore's scoreList().  QtConcurrent seems to know my CPU count (I'm on a 12-core machine) and automatically create that many threads without me having to specify that, which is nice. 

Anyway, here are some speed measurements on MSVC release build (using the Thunderbolt_March.mscz attached to that "Score Editing Slow" issue mentioned earlier) by putting timers at start and end of Score::update():

- first update (entire range) upon initial load of score: 3701 ms single threaded becomes 1360 ms with concurrent update, almost 3 times faster
- an incremental update of just changing first pitch of 2nd staff 4th meas in main score: 1114 ms becomes 870 ms, only ~28% faster.

So it seems the biggest improvement comes from updating the entire score.  That is because every part score has to relayout it's entire score.  But not much speedup when just change single note.  And that is because the main score has a lot (34) staves.  So even though the system that is being updated is only 8 measures long, there are so many staves that it takes much longer.  Compared to every individual part score where the system to be regenerated has only one staff (or 2 in case of grand-staff instruments).  But still it is nice to have all these part scores do update concurrently, so user doesn't have to wait for all of the easier part scores to finish, (even though they're easier, there are so many that their time does add up).

Here's a screen shot of the CPU usage...100% means that all 12 cores are being utilized, and you can see that does indeed happen with that biggest spike where, which occurs when doing relayout of entire score (which is the case upon initial load of a score or doing any operation which mandates full relayout).  The subsequent smaller spikes are when I change a single pitch...and unfortunately only has a much smaller peak...but at least it is some speedup:

![concurrent_release-mode](https://user-images.githubusercontent.com/6502474/50631375-09d08800-0f12-11e9-8359-b7220856d026.PNG)

Anyway, I think the takeaway point is that layout speed for main scores with very large systems still needs to be significantly improved...there is probably a lot of redundant computations...as I'm sure most people are aware of.  Also as I mentioned in pictures in the original issue report, the hottest line is calculating bezier rectangles...which surely could be simplified...

* Note: layout isn't quite completely indpendent...there are some exceptions which can be fixed...such as initBeamMetrics which causes sporadic crash on first concurrent update, I think because two threads might try to do that init at the same time...obviously that initBeamMetrics should be moved out of the parallel section.  If there are any other non-independent parts in layout, I think they can also be made independent.

ericfont commented:
note, this is not an actual PR...this is just my own investigation that I'm sharing for people to comment.

Jojo-Schmitz commented on libmscore/cmd.cpp:
```diff
@@ -204,41 +204,67 @@ void CmdState::dump()
 #endif
 
+//---------------------------------------------------------
+//   update
```
wrong (copy/paste) comment

ericfont replied:
thanks.  I've fixed that...will be included when I next push.

Jojo-Schmitz commented on libmscore/score.h:
```diff
@@ -997,7 +1000,7 @@ class Score : public QObject, public ScoreElement {
 
       void doLayout();
-      void doLayoutRange(int, int);
+      void doLayoutRange(int stick, int etick);
```
parameter names not needed

ericfont replied:
well I realize that...to me including the parameters is nicer, but I'll remove that from now on.

anatoly-os replied:
In this particular case I would keep the names. The reason is simple, when you want to use this or that method, you look at the declaration, not the definition. Two int parameters may mean everything. But stick, etick make things clear to the reader.

You don't need to specify names if parameter type is obvious (e.g. Score*), but in this case I strongly recommend to keep the names.

ericfont commented:
(note...travis mtest build will fail, because I have to do something to make it build properly on linux.

ericfont commented on libmscore/cmd.cpp:
```diff
@@ -204,41 +204,67 @@ void CmdState::dump()
 //    layout & update
 //---------------------------------------------------------
 
 void Score::update()
```
(as an aside, in my opinion, the name of the main function "Score::update()" is a little confusing, because it actually acts upon all scores in the MasterScore, and it does both layout (if necessary) and repaint (if necessary)...I think there is a better name for this, but I don't think I have permission to change it.)

ericfont commented:
As I was mentioning, initBeamMetrics (https://github.com/musescore/MuseScore/blob/26d15105ab26c47c5866621051d6b025d4a28d36/libmscore/beam.cpp#L676) causes crash due to race condition on first load...so I think if I moved that init stuff  outside of layout call path will work.  FYI: here is the stack trace from that crash: https://gist.github.com/ericfont/681865eeba1ab2d0e06d0bb5e5b5bc8b

ericfont commented:
Just some more stats running on my 4-core x86-64 linux machine now (debug build, unfortunately I haven't build release):
- Initial full layout & update: 10627 ms single-threaded gets sped up to 5102 ms...about twice as fast 
- Changing one note: 2855 ms single-threaded becomes 2035...about %40 speedup

ericfont commented:
Ok, I think I fixed the initBeamMetrics race conditions using a simple QMutex with a tryLock.

And I think I fixed the Travis linux compiler error by converting qCDebug statements from being like printf statements into being like C++ cout-style statements (I don't know what that works...according to Qt docs, both styles work).

ericfont commented:
I can't figure out why Travis is unable to build with the qCDebug statements, either in printf format or in regular format...they're perfectly valid https://doc.qt.io/qt-5/qloggingcategory.html#qCDebug and I defined the logging categories.  It builds fine & passes tests on both my windows MSVC and linux QtCreator machines.

Justin135 commented:
Maybe you could also add a simple script that updates only the parts that have the edited instrument. So, instead of updating all of the parts in the background, only the one part needs to be updated.

Justin135 commented:
I would help editing the code, but I'm not familiar with C++. I do know Java, and I've been thinking about learning C++, so I feel like it will make learning the syntax easier.

ericfont commented:
@Justin135, I'm not going got try to make too many changes in the PR because I'm keeping it focused just on making multithreaded.

the 3.0 code already does try to do a related optimization in this if block for "update": https://github.com/musescore/MuseScore/blob/e2950adf4ce07cfb8291b6404967ca9766455dd1/libmscore/cmd.cpp#L235.  But it should be noted that the term "update" here has a specific meaning...it means "redraw".  The code tries to be smart by only doing redraws for the current screen for the refresh rectangle of the part that changed.  But it doesn't try limiting the relayout in the main score to be only those instruments.

I've also been noticing that a lot of Score::setLayoutAll() calls seem excessive & unnecessary to me (https://musescore.org/en/node/281091), but again that is a separate issue.

But really quite a lot of changes will have to be made to do what you said.  First the Score::setLayout(int tick) function will have to have an option to only mark layout as dirty for a specific instrument.  So then would have to keep track of dirty or clean state for every measure for each staff.  And then would have to rework the doLayoutRange code such that it can reuse old clean staves and measures.  That will be quite a more involved fix then merely "add a simple script".

ericfont commented:
I'm debugging the crash from the failed test (exchange-voice-part.script) and notice it is because inside Score:: doLayoutRange in call for getNextMeasure i does Chord:: layoutStem1 which has to "Create stem if necessary", and that stem is put onto the *master score's* undostack...  So that breaks the thread safeness.  Maybe putting a QMutex to lock undoAddElement for each score will do the trick.

ericfont commented:
I think I got tests working now...turns out that Scores may modify the MasterScore's UndoStack during layout, so in order to allow multiple concurrent scores doing layout, I put a QMutex around every UndoStack function which modified the UndoStack.  If each score had it's own undostack for the purposes of layout, then I wouldn't have to do this.

Jojo-Schmitz commented on libmscore/undo.cpp:
```diff
@@ -318,6 +331,7 @@ void UndoStack::rollback()
+      mutexPushPop.lock();
       if (curCmd == 0) {
             qWarning("not active");
             return;
```
unlock missing here

Jojo-Schmitz replied:
or just lock only after this early exit

ericfont replied:
you are right...I missed that.  Actually Joshua just informed me of a useful convenience class QMutexLocker (https://doc.qt.io/qt-5/qmutexlocker.html) which does exactly what I need here and avoids such mistakes.

Jojo-Schmitz replied:
Ah, that'd make it far easier indeed.
Should probably get used in some more places thoughout the MuseScore code

Jojo-Schmitz commented on libmscore/undo.cpp:
```diff
@@ -209,11 +209,13 @@ UndoStack::~UndoStack()
+      mutexPushPop.lock();
       if (curCmd) {
             qWarning("already active");
             return;
```
unlock missing. Or move the lock to after this early exit

dmitrio95 commented:
Archiving this pull request
